testnet-contracts/contracts/BridgeBank/BridgeBank.sol

```diff
@@ -0,0 +1,271 @@
+    *       This feature is used for testing and is available at the operator's own risk.
+    */
+    function() external payable onlyOperator {}
+
+    /*
+    * @dev: Creates a new BridgeToken
+    *
+    * @param _symbol: The new BridgeToken's symbol
+    * @return: The new BridgeToken contract's address
+    */
+    function createNewBridgeToken(
+        string memory _symbol
+    )
+        public
+        onlyOperator
+        returns(address)
+    {
+        return deployNewBridgeToken(_symbol);
+    }
+
+    /*
+     * @dev: Mints new BankTokens
+     *
+     * @param _cosmosSender: The sender's Cosmos address in bytes.
+     * @param _ethereumRecipient: The intended recipient's Ethereum address.
+     * @param _cosmosTokenAddress: The currency type
+     * @param _symbol: comsos token symbol
+     * @param _amount: number of comsos tokens to be minted
+\    */
+     function mintBridgeTokens(
+        bytes memory _cosmosSender,
+        address payable _intendedRecipient,
+        address _bridgeTokenAddress,
+        string memory _symbol,
+        uint256 _amount
+    )
+        public
+        onlyCosmosBridge
+    {
+        return mintNewBridgeTokens(
+            _cosmosSender,
+            _intendedRecipient,
+            _bridgeTokenAddress,
+            _symbol,
+            _amount
+        );
+    }
+
+    /*
+     * @dev: Locks received Ethereum funds.
+     *
+     * @param _recipient: bytes representation of destination address.
+     * @param _token: token address in origin chain (0x0 if ethereum)
+     * @param _amount: value of deposit
+     */
+    function lock(
+        bytes memory _recipient,
+        address _token,
+        uint256 _amount
+    )
+        public
+        payable
+        availableNonce
+        returns(bytes32 _id)
+    {
+        string memory symbol;
+
+        // Ethereum deposit
+        if (msg.value > 0) {
+          require(
+              _token == address(0),
+              "Ethereum deposits require the 'token' address to be the null address"
+            );
+          require(
+              msg.value == _amount,
+              "The transactions value must be equal the specified amount (in wei)"
+            );
+
+          // Set the the symbol to ETH
+          symbol = "ETH";
+          // ERC20 deposit
+        } else {
+          require(
+              BridgeToken(_token).transferFrom(msg.sender, address(this), _amount),
+              "Contract token allowances insufficient to complete this lock request"
+          );
+          // Set symbol to the ERC20 token's symbol
+          symbol = BridgeToken(_token).symbol();
+        }
+
+        return newEthereumDeposit(
+            msg.sender,
+            _recipient,
+            _token,
+            symbol,
+            _amount
+        );
+    }
+
+    /*
+     * @dev: Unlocks Ethereum deposits.
+     *
+     * @param _id: Unique key of the CosmosDeposit.
+     */
+    function unlock(
```

Comment:
is this function just for the operator, as a backdoor? its not really needed/part of the end-goal production version right?

it also looks like there is no implementation yet of unlocking ethereum/erc20 tokens to be sent out to a receiver under the typical bridge flow of claims/prophecies - is that missing? specifically, i mean the flow of burning cETH on cosmos with a target destination ethereum address, then waiting for relayers/validators to make the claims and final prophecy that unlocks that real ETH and sends it to the destination chosen during the initial burn.

maybe these two flows are being confused here

Reply:
@denalimarsh bump?

Reply:
@musnit after cETH is burned on Cosmos, details of the burn event are relayed to CosmosBridge's `newProphecyClaim()` with requests for the final prophecy sent to Oracle's `processBridgeProphecy()`. If the final prophecy is successful then BridgeBank will mint tokens to the intended recipient address. This is done independently of contract funds that were on-boarded from the `lock()` function. The tokens minted by BridgeBank upon the successful processing of the final prophecy are BridgeTokens, tokens which represent the Cosmos based assets 1:1. Ethereum is also currently processed by this system, with Ethereum represented as a BridgeToken with token contract address `address(0)`.

Given this system, I propose that we update the `unlock()` function so that it swaps the sender's BridgeTokens for real Ethereum/ERC20 and burns the received BridgeTokens. If someone is holding BridgeTokens they can't redeem because nobody has locked the correlated ERC20 token on the Bridge, the BridgeTokens still have the same value since they remain redeemable 1:1. This also allows for the burning of Cosmos assets -> minting of BridgeTokens even if Bridge does not hold a balance of the ERC20 token type that the BridgeTokens represent at the time of the burn event.

Reply:
"burning of Cosmos assets -> minting of BridgeTokens even if Bridge does not hold a balance of the ERC20 token type that the BridgeTokens represent at the time of the burn event" -> this would only happen if the bridge broke or the validators maliciously attacked and withdrew those assets when they weren't supposed to?

Reply:
i suppose ideally as soon as cETH is burned, that swap should happen immediately and so effectively the whole burn ->mint -> swap -> burn flow becomes equivalent to the burn -> unlock flow, and the latter seems much simpler?

Reply:
Is it the case that assets can only be originally on-boarded to the Bridge from Ethereum? I was under the assumption that Cosmos could also be the starting point for new funds.

Reply:
with the burn->unlock flow, yes.

there is also a lock-> mint flow from cosmos to eth with lockup ATOM on cosmos, mint eATOM that would support the case you mention

Reply:
Okay, that is indeed much simpler. I will update the contracts to reflect those asset flows :)

Reply:
https://github.com/cosmos/peggy/pull/82/commits/159cf28ee1ba897533c941aa2dfe0bbd993f29a3 includes the updated contracts and test suite